### ISSUE TITLE
Decoupled logic cannot be tested via Spring MockMvc

### ISSUE THREAD

**Issue by djkeh** [thymeleaf/thymeleaf-spring#152]
Hi,
I encountered a problem when I tried to decouple the template logic and test using Spring MockMvc. Here's the detail.

##### Environment

  * Version of Thymeleaf: 3.0.7
  * Environment:
    * IDE: Eclipse Oxygen (4.7)
    * Spring: 4.3.10
    * Spring Boot: 1.5.6
    * Twitter Bootstrap: 3.3.7

##### How to reproduce

These are parts of decoupled template code.

```html
<!-- target (index.html) -->

<nav class="navbar navbar-inverse navbar-fixed-top">
  <div class="navbar-header">
    <a class="navbar-brand" id="th-navbar-brand" href="/">Brand Logo</a>
  </div>
</nav>

<div class="container">
  <div class="row" id="th-shortened-url">
    <div class="col-md-2">
      <h4>URL:</h4>
    </div>
    <div class="col-md-10">
      <div class="well"></div>
    </div>
  </div>
</div> <!-- /container -->
```

```xml
<!-- decoupled logic (index.th.xml) -->

<?xml version="1.0"?>
<thlogic>
  <attr sel="#th-navbar-brand" th:text="'Testing decoupled logic'" />
  <attr sel="#th-shortened-url" th:if="${url}">
    <attr sel="/div[1]">
      <attr sel="/.well" th:text="${url}" />
    </attr>
  </attr>
</thlogic>
```

These codes above actually works pretty well in the browser as intended.

* `navbar-brand` element has a proper text: `Testing decoupled logic`
* If the attribute `url` exists, template draws `th-shortened-url` element and its contents.
* If does not, template prevents drawing `th-shortened-url` element.

This is the test code.

```java
import static org.hamcrest.CoreMatchers.containsString;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;

@RunWith(SpringRunner.class)
@WebMvcTest(MainController.class)
public class MainControllerTest {
    
    @Autowired
    private MockMvc mockMvc;

    @Test
    public void testRequest() throws Exception {
        mockMvc.perform(post("/").param("url", "http://test.url"))
                .andDo(print())
                .andExpect(status().isOk())
                .andExpect(content().string(containsString("Testing decoupled logic")));
                .andExpect(content().string(containsString("http://test.url")));
    }
}
```

This test code runs Spring application without running server.
We can see the result output through the code `.andDo(print())`,
and this is the printed MockMvc result.

```html
<nav class="navbar navbar-inverse navbar-fixed-top">
  <div class="navbar-header">
    <a class="navbar-brand" id="th-navbar-brand" href="/">Brand Logo</a>
  </div>
</nav>

<div class="container">
    <div class="row" id="th-shortened-url">
      <div class="col-md-2">
        <h4>Shortened URL:</h4>
      </div>
      <div class="col-md-10">
        <div class="well"></div>
      </div>
    </div>
    
  </div> <!-- /container -->
```

As you can see,

* `Brand Logo` isn't replaced.
* `well` class in `th-shortened-url` element doesn't have expected value `http://test.url`.

Additionally, if I `perform()` request without `url` parameter, the html element `th-shortened-url` is not supposed to be drawn because of `th:if` statement, but it still remains in the response.

##### Possible workarounds

I just abandoned decoupling technique and put thymeleaf attributes directly into the original html file just like the conventional way. That works.

```html
<!-- html -->

<nav class="navbar navbar-inverse navbar-fixed-top">
  <div class="navbar-header">
    <a class="navbar-brand" id="th-navbar-brand" href="/" th:text="'Testing decoupled logic'">Brand Logo</a>
  </div>
</nav>
```

```html
<!-- MockMvc result -->

 <nav class="navbar navbar-inverse navbar-fixed-top">
  <div class="navbar-header">
    <a class="navbar-brand" id="th-navbar-brand" href="/">Testing decoupled logic</a>
  </div>
</nav>
```

I think this clearly shows that thymeleaf engine is working through the MockMvc process, but the decoupling feature is not activating correctly.

Thanks.

**Comment by danielfernandez** [comment]
I lack deep knowledge on how the Spring MockMvc mechanism works… could you please give me more detail on how this is an issue at the Thymeleaf side and not MockMvc itself or the configuration of the Thymeleaf engine performed by the latter?

**Comment by djkeh** [comment]
@danielfernandez 

I'm afraid I can't explain the detailed MockMvc mechanism... but I think I can explain its basic concept and the reference.

* MockMvc: Main entry point for server-side Spring MVC test support. The main purpose of using this object is to **test web application without starting the actual server.** According to the reference, the codes are called _exactly the same way_ as if it was processing a real HTTP request.
* The official tutorial has more detailed explanation with example codes.
  * https://spring.io/guides/gs/testing-web/

And Here are some results I figured out.

### Check 1 : MockMvc returns well processed basic Thymeleaf template.

As the above I showed in the `Possible workarounds` section, MockMvc seems to process Thymeleaf attributes in the template.

### Check 2 : MockMvc returns **well processed decoupling template as well in `@SpringBootTest` annotated test.**

I just confirmed that this code works:

```java
// Test Approach 1
@RunWith(SpringRunner.class)
@SpringBootTest
@AutoConfigureMockMvc
public class Thymeleaf3Test {
    @Autowired
    private MockMvc mvc;
    
    @Test
    public void test() throws Exception {
        mvc.perform(get("/"))
                .andDo(print()) // prints MockMvc response result.
                .andExpect(status().isOk()) // checks http status code of MockMvc response.
                .andExpect(content().string(containsString("decoupling logic test"))); // checks string pattern match: "decoupling logic test"
    }
}
```

This code below still doesn't work.

```java
// Test Approach 2
@RunWith(SpringRunner.class)
@WebMvcTest(MainController.class)
public class MainControllerTest {
    
    @Autowired
    private MockMvc mvc;
    
    @Test
    public void test() throws Exception {
        mvc.perform(get("/"))
                .andDo(print())
                .andExpect(status().isOk())
                .andExpect(content().string(containsString("decoupling logic test")));
    }
}
```

So it seems to be the problem of Spring test annotation. The only difference is `@SpringBootTest + @AutoConfigureMockMvc` and `@WebMvcTest`. What is that?

* `@SpringBootTest + @AutoConfigureMockMvc`: the full Spring application context is started. Server is not running.
* `@WebMvcTest`: the web layer only is instantiated by the Spring Boot. Server is not running as well.

### Check 3 : The problem happens when the test is performed with `@WebMvcTest`

In the code above, `Test Approach 1(@SpringBootTest)` handles not only Thymeleaf attributes in the code but also its decoupled logic codes. On the other hand, the code `Test Approach 2(@WebMvcTest)` only shows the right result with Thymeleaf attributes, not with the decoupled codes. I think there's something to do with this point.

### Check 4 : Thymeleaf 3 customized options:

Here's it is.

```yml
spring:
  thymeleaf:
    decoupled-logic: true
```

**Comment by danielfernandez** [comment]
I'm a bit confused by your `spring.thymeleaf.decoupled-logic` configuration property…

As far as I know, this is not supported by Spring Boot (nor `1.5.6.RELEASE` nor `2.0.0.BUILD-SNAPSHOT`). Is this something you are adding manually somehow to the default allowed properties (`ThymeleafProperties` class) in Spring Boot? If so, this could very well the reason for your differences, as `@WebMvcTest` will not initialise your entire Spring Boot infrastructure.

**Comment by djkeh** [comment]
@danielfernandez 

Oh... That's correct. I've been using customized configuration property to use decoupled logic feature in Spring Boot `1.5.6` because Thymeleaf 3 and its proper auto-configuration are not yet included in the framework. I forgot to explain this. Here's the full gist code for using my customized configuration property.

https://gist.github.com/djkeh/dce08aa2b66b52bd56cdc1105eb168c1

You're right. `@WebMvcTest` is only meant to scan `@Controller`, `@ControllerAdvice`, `@JsonComponent`, `Filter`, `WebMvcConfigurer` and `HandlerMethodArgumentResolver` beans, so `@Component` beans are not supposed to be scanned, which seemed to make this issue happen. I just found related documents here:

https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-testing.html#boot-features-testing-spring-boot-applications-testing-autoconfigured-mvc-tests

What I understand is that Thymeleaf 3 is coming into the Spring Boot 2 in future. `@WebMvcTest` is supposed to use `ThymeleafAutoConfiguration` included in the framework, so this problem would only be temporary. I think now there will be some workarounds to solve this issue:

#### 1. Use `@SpringBootTest` instead of `@WebMvcTest`

This definitely works. The tradeoff is to instantiate whole Spring context instead of instantiating web layer only. The test speed can be a little bit slower.

#### 2. Abandon decoupled logic

Works perfectly as well but not so desirable.

#### 3. Load custom Thymeleaf 3 auto-configuration bean while testing with `@WebMvcTest`

This is of my concern, but I couldn't figure out how to do it. **Please help if you know how to do this.**

#### 4. Wait for Spring Boot 2

This problem is expected to be naturally solved after Spring Boot is updated to version 2.

**Comment by danielfernandez** [comment]
> This problem is expected to be naturally solved after Spring Boot is updated to version 2.

Actually, I've confirmed that the `ThymeleafProperties` class for the current Spring Boot `2.0.0.BUILD-SNAPSHOT` does not contain any properties for enabling decoupled logic, so if this is something you'd like to see in Spring Boot 2, I would suggest adding an enhancement ticket there.

So I believe we can close this ticket now, right?

**Comment by djkeh** [comment]
@danielfernandez Yes. Even if the new properties are not currently included in the `2.0.0.BUILD-SNAPSHOT`, I think they're aware of it. I'll create an enhancement ticket though.

I believe the workaround **no.3** I mentioned in the comment before might be the best solution for now in Spring version 1, if there's a way how to achieve it. So after this issue is closed, I'm still expecting someone who knows would come and leave a comment for it.

##### summerize

It seems like we're currently not able to test Thymeleaf decoupled-logic template using only `@WebMvcTest` under Spring Boot 1. We can still test it by using `@SpringBootTest(webEnvironment = WebEnvironment.NONE)` instead, which would work efficiently enough on configuration tests. If anyone know better way to test user-customized configuration, it'd be very appreciated to share.

We can close this issue. Thank you for your consideration.
